### PR TITLE
Make Unique family of APIs use reduce_then_scan

### PR DESCRIPTION
**Pull request by danhoeflinger**
This PR changes the unique family of scan-like APIs to use reduce_then_scan when it is beneficial.

This PR allows us to remove `__pattern_scan_copy` functions because they are no longer used.  
We have moved the algorithm decisions to go through at the level of `__parallel_[copy_if/partition/unique]_copy`, so we no longer need the "scan_copy" at the pattern level.

Moves all algorithm selection decisions to `__parallel_unique_copy`, and unifies range API to also use this function. This allows us to unify the algorithmic selection, and provide performance improvements to the ranges API.

Unique requires some `constexpr` special casing in the kernel to allow us to avoid an extra branch for each element in the `_GenMask` for unique to avoid underflow when `index == 0`.  We special case the kernels to skip and always copy the 0th element for unique family APIs, and start the scan at element `1`.  This allows us to handle this copy of the 0th element without any additional kernel launches.  The `n==1` case is handled specially with a simple copy call.

---

This PR is targeted to #1764, to allow for a clean diff, and is a part of the following sequence of PRs meant to be merged in order:

#1769 [MERGED] ~~Relocate __lazy_ctor_storage to utils header~~
#1770 [MERGED] ~~Use __result_and_scratch_storage within scan kernels~~
#1762 Add reduce_then_scan algorithm for transform scan family
#1763 Make Copy_if family of APIs use reduce_then_scan algorithm
#1764 Make Partition family of APIs use reduce_then_scan algorithm 
#1765 Make Unique family of APIs use reduce_then_scan algorithm (This PR)

This work is a collaboration between @mmichel11 @adamfidel and @danhoeflinger, and based upon an original prototype by Ted Painter.

### CODE REVIEW

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:390]:
```suggestion
        views::all_write(std::forward<_Range2>(__result)), __pred);
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:404]:
```suggestion
                std::forward<_ExecutionPolicy>(__exec)),
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:442]:
```suggestion
                                   std::forward<_ExecutionPolicy>(__exec)),
```

File: include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h:389]:
```suggestion
        __dispatch_tag, std::forward<_ExecutionPolicy>(__exec), views::all_read(std::forward<_Range1>(__rng)),
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:444]:
```suggestion
                               std::forward<_Range>(__rng));
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:400]:
```suggestion
        using _CopyBrick = oneapi::dpl::__internal::__brick_copy<__hetero_tag<_BackendTag>, _ExecutionPolicy>;
```

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h]: fixed, thanks.

**Review by julianmi — APPROVED** [review]
LGTM. I would hold off merging this until the main patch was merged.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:968]:
```suggestion
        // For a sequence of size 1, we can just copy the only element to the result.
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:400]: could use a similar comment to above:
`// For a sequence of size 1, we can just copy the only element to the result.`

**Review by danhoeflinger — COMMENTED** [review]
LGTM after rebasing and minor comment changes.

**Review by julianmi — DISMISSED** [review]
LGTM after a rebase.

**Review by mmichel11 — APPROVED** [review]
LGTM

**Pull request merged**
